openeo/extra/artifacts/internal_s3/artifact_helper.py

```diff
@@ -0,0 +1,68 @@
+from __future__ import annotations
+
+import datetime
+from typing import TYPE_CHECKING
+from boto3.s3.transfer import TransferConfig
+
+from openeo.extra.artifacts.artifact_helper import ArtifactHelper
+from openeo.extra.artifacts.internal_s3.sts import OpenEOSTSClient
+from openeo.extra.artifacts.internal_s3.config import S3Config
+from openeo.extra.artifacts.internal_s3.model import S3URI
+
+if TYPE_CHECKING:
+    from pathlib import Path
+    from openeo.rest.connection import Connection
+    from openeo.extra.artifacts.internal_s3.model import AWSSTSCredentials
+
+
+class S3ArtifactHelper(ArtifactHelper):
+    BUCKET_NAME = "openeo-artifacts"
+    # From what size will we switch to multi-part-upload
+    MULTIPART_THRESHOLD_IN_MB = 50
+
+    def __init__(self, creds: AWSSTSCredentials, config: S3Config):
+        super().__init__(config)
+        self._creds = creds
+        self.s3 = config.build_client("s3", session_kwargs=creds.as_kwargs())
+    
+    @classmethod
+    def _from_openeo_connection(cls, conn: Connection, config: S3Config) -> S3ArtifactHelper:
+        sts = OpenEOSTSClient(config=config)
+        creds = sts.assume_from_openeo_connection(conn)
+        return S3ArtifactHelper(creds, config=config)
+
+    def _user_prefix(self) -> str:
+        """Each user has its own prefix retrieve it"""
+        return self._creds.get_user_hash()
+    
+    def _get_upload_prefix(self) -> str:
+        return f"{self._user_prefix()}/{datetime.datetime.now(datetime.UTC).strftime('%Y/%m/%d')}/"
+    
+    def _get_upload_key(self, object_name: str) -> str:
+        return f"{self._get_upload_prefix()}{object_name}"
+    
+    def upload_file(self, object_name: str, src_file_path: str | Path) -> S3URI:
```

Comment:
Isn't it more logical to put `src_file_path` argument before `object_name` (like bash `cp src dst`).
It would also allow to make `object_name` optional  provide a autogenerated name.

minor note: `src_file_path` looks a bit long as name, could be just `path`, `source` or `src` I think

Reply:
Yes makes a lot of sense as object name naturally can be the filename by default and only special cases would want to chose the target object name.

Incorporated feedback including minor note.